dataclients/kubernetes/clusterclient.go

```diff
@@ -0,0 +1,374 @@
+	if err != nil {
+		return nil, err
+	}
+
+	httpClient, err := buildHTTPClient(serviceAccountDir+serviceAccountRootCAKey, o.KubernetesInCluster, quit)
+	if err != nil {
+		return nil, err
+	}
+
+	ingClsRx, err := regexp.Compile(ingCls)
+	if err != nil {
+		return nil, err
+	}
+
+	c := &clusterClient{
+		ingressesURI:   ingressesClusterURI,
+		routeGroupsURI: routeGroupsClusterURI,
+		servicesURI:    servicesClusterURI,
+		endpointsURI:   endpointsClusterURI,
+		ingressClass:   ingClsRx,
+		httpClient:     httpClient,
+		token:          token,
+		apiURL:         apiURL,
+	}
+
+	if o.KubernetesNamespace != "" {
+		c.setNamespace(o.KubernetesNamespace)
+	}
+
+	return c, nil
+}
+
+func (c *clusterClient) setNamespace(namespace string) {
+	c.ingressesURI = fmt.Sprintf(ingressesNamespaceFmt, namespace)
+	c.routeGroupsURI = fmt.Sprintf(routeGroupsNamespaceFmt, namespace)
+	c.servicesURI = fmt.Sprintf(servicesNamespaceFmt, namespace)
+	c.endpointsURI = fmt.Sprintf(endpointsNamespaceFmt, namespace)
+}
+
+func (c *clusterClient) createRequest(uri string, body io.Reader) (*http.Request, error) {
+	req, err := http.NewRequest("GET", c.apiURL+uri, body)
+	if err != nil {
+		return nil, err
+	}
+
+	if c.token != "" {
+		req.Header.Set("Authorization", "Bearer "+c.token)
+	}
+
+	return req, nil
+}
+
+func (c *clusterClient) getJSON(uri string, a interface{}) error {
+	log.Debugf("making request to: %s", uri)
+
+	req, err := c.createRequest(uri, nil)
+	if err != nil {
+		return err
+	}
+
+	rsp, err := c.httpClient.Do(req)
+	if err != nil {
+		log.Debugf("request to %s failed: %v", uri, err)
+		return err
+	}
+
+	log.Debugf("request to %s succeeded", uri)
+	defer rsp.Body.Close()
+
+	if rsp.StatusCode == http.StatusNotFound {
+		return errResourceNotFound
+	}
+
+	if rsp.StatusCode != http.StatusOK {
+		log.Debugf("request failed, status: %d, %s", rsp.StatusCode, rsp.Status)
+		return fmt.Errorf("request failed, status: %d, %s", rsp.StatusCode, rsp.Status)
+	}
+
+	b := bytes.NewBuffer(nil)
+	if _, err = io.Copy(b, rsp.Body); err != nil {
+		log.Debugf("reading response body failed: %v", err)
+		return err
+	}
+
+	err = json.Unmarshal(b.Bytes(), a)
+	if err != nil {
+		log.Debugf("invalid response format: %v", err)
+	}
+
+	return err
+}
+
+func (c *clusterClient) clusterHasRouteGroups() (bool, error) {
+	var crl clusterResourceList
+	if err := c.getJSON(clusterZalandoResourcesURI, &crl); err != nil {
+		return false, err
+	}
+
+	for _, cr := range crl.Items {
+		if cr.Name == routeGroupsName {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// filterIngressesByClass will filter only the ingresses that have the valid class, these are
+// the defined one, empty string class or not class at all
+func (c *clusterClient) filterIngressesByClass(items []*ingressItem) []*ingressItem {
+	validIngs := []*ingressItem{}
+
+	for _, ing := range items {
+		// No metadata is the same as no annotations for us
+		if ing.Metadata != nil {
```

Comment:
Aren't we already sure we have only valid ingress objects?

Reply:
the point is that it is coming from an external process, even if we expect it to not contain bugs, we cannot be sure, and we don't want to crash our process.

Note that in the case of the route gropus, we are adding now full validation in a single step, but that also comes only after the cluster client has loaded the current route groups.

Reply:
I realized it later, that this is basically the earliest validation